core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultLockManager.java

```diff
@@ -346,6 +358,11 @@ public void addListener(LockListener listener) {
       }
 
       @Override
+      public CompletableFuture<LockPromise> asCompletableFuture() {
+         return notifier.thenApplyAsync(lockState1 -> CompositeLockPromise.this);
```

Comment:
It deadlocks if I don't use async. What is the issue with the common pool?

Reply:
We don't want the common pool to later then block on something. We should always make sure it only ever can run CPU intensive code. I am not familiar with what else is down the stack, but if everything else supports async, it should technically be fine.

Reply:
it can be anything (including user code).
Also it can increase the latency of write operation when the primary owner releases the lock, it starts to run the next command instead of continue handling the rest of the interceptors and reply to the originator. 
this is also a problem for the schedule executor since it can may the only thread executing "normal" code instead of handling the scheduled tasks.